Feature/6 create power fault

deltaMinor opened the pull request:
The power fault scene is directly added to the game scene temporarily until the spawning system is implemented

deltaMinor committed:
Feature/6 create power fault

The power fault scene is directly added to the game scene temporarily until the spawning system is implemented

Cxyu629 commented on scripts/enums.gd:
```diff
@@ -5,4 +5,6 @@ enum Team {
 	PLAYER,
 	ENEMY,
 	TURRET,
+	SABOTAGED_FAULT,
```
What's the rationale behind having two separate teams? @deltaMinor

deltaMinor replied:
The intended default targeting behaviours for the current entities are as follows:

| Target | Player | Turret | Enemy |
| ------------- | ------------- | ------------- | ------------- |
| Player | Depends on multiplayer implementation* | Do not target | Always target |
| Turret | Target with RepairComp | Do not target | Always target |
| Enemy | Target with MeleeComp | Always target | Do not target |
| Sabotaged Fault | Always target | Do not target | Do not target |
| Rebooting Fault | Not relevant* | Do not target | Always target | 

*May be subject to changes
Thus, perhaps the fault can switch to the Turret's team when rebooting, however it will need to switch between three different teams(for sabotaged, rebooting and NONE when it's fixed) regardless. I initially avoided having the Turret team include a rebooting fault for the sake of clarity, however I don't think there would be any issues in doing so, if the team name is renamed to something like "BUILDINGS"(or some other name that indicates structures belonging to the player), should I go ahead with that instead? @Cxyu629

Cxyu629 replied:
At this point it sounds like the team implementation should behave more like the collision layers and masks, meaning each entity can belong to multiple teams.

It's fine to leave it as it is for the moment. Maybe revisit this when the team implementation is modified in the future.

Cxyu629 commented on scenes/fault.tscn:
```diff
@@ -0,0 +1,526 @@
+"transitions": PackedFloat32Array(1),
+"values": [{
+"args": [],
+"method": &"exit"
```
```suggestion
"method": &"emit_disappear_finished"
```

Cxyu629 commented on scenes/fault.tscn:
```diff
@@ -0,0 +1,526 @@
+[node name="RebootTimer" type="Timer" parent="."]
+wait_time = 10.0
+
+[connection signal="disappeared" from="Visuals" to="." method="_on_visuals_disappeared"]
```
```suggestion
[connection signal="disappear_finished" from="Visuals" to="." method="_on_visuals_disappear_finished"]
```

Cxyu629 commented on scripts/fault_scene/visuals.gd:
```diff
@@ -0,0 +1,62 @@
+
+func _process(delta: float) -> void:
+	if fault.state == Fault.State.REBOOTING:
+		rebooting_bar.value = fault.get_time_progress_ratio() * 100
```
```suggestion

```

Cxyu629 commented on scripts/fault_scene/fault.gd:
```diff
@@ -0,0 +1,64 @@
+
+
+
+func _on_visuals_disappeared() -> void:
```
```suggestion
func _on_visuals_disappear_finished() -> void:
```

Cxyu629 commented on scripts/player_scene/visuals.gd:
```diff
@@ -58,22 +59,14 @@ func play_gun_fire(_bullet):
-
-
 func _on_gun_sprite_animation_finished() -> void:
 	if (gun_sprite.animation == "melee_fire"):
```
```suggestion
	if gun_sprite.animation == "melee_fire":
```

Cxyu629 commented on scripts/fault_scene/fault.gd:
```diff
@@ -0,0 +1,64 @@
+
+@export var repair_target : float
+@export var repair_prop : RepairProp
+@export var hitbox : HitboxComp
```
```suggestion
@export var repair_target: float
@export var repair_prop: RepairProp
@export var hitbox: HitboxComp
```

Cxyu629 commented on scripts/fault_scene/fault.gd:
```diff
@@ -0,0 +1,64 @@
+			handle_state_changed(prev_state, state)
+			state_changed.emit(prev_state, state)
+
+signal state_changed(from: State, to: State)
```
```suggestion
@onready var reboot_timer := $RebootTimer as Timer

```

Cxyu629 commented on scripts/fault_scene/visuals.gd:
```diff
@@ -0,0 +1,62 @@
+
+func update_repair_progress(progress: float):
+	repair_bar.value = progress * 100
+	if(repair_bar.value > 0):
```
```suggestion
	if repair_bar.value > 0:
```

Cxyu629 commented on scripts/fault_scene/fault.gd:
```diff
@@ -0,0 +1,64 @@
+
+func check_repair(value: float) -> void:
+	if state == State.SABOTAGED:
+		if(value >= repair_target):
```
```suggestion
		if value >= repair_target:
```

Cxyu629 commented on scripts/fault_scene/visuals.gd:
```diff
@@ -0,0 +1,62 @@
+	repair_bar.visible = false
+	rebooting_bar.visible = false
+	fault.state_changed.connect(handle_state_changed)
+	fault.repair_progress.connect(update_repair_progress)
```
```suggestion
	fault.repair_progressed.connect(update_repair_progress)


func _process(_delta: float) -> void:
	if fault.state == Fault.State.REBOOTING:
		rebooting_bar.value = fault.get_time_progress_ratio() * 100


func handle_state_changed(from: Turret.State, to: Turret.State) -> void:
	match [from, to]:
		[Fault.State.SABOTAGED, Fault.State.REBOOTING]:
			play_rebooting()
		
		[Fault.State.REBOOTING, Fault.State.SABOTAGED]:
			play_sabotaged()
		
		[Fault.State.REBOOTING, Fault.State.FIXED]:
			play_fixed()

```

Cxyu629 commented on scripts/properties/repair_prop.gd:
```diff
@@ -0,0 +1,22 @@
+	if value != _prev_value:
+		just_changed.emit(value)
+	_prev_value = value
+
```
```suggestion


```

Cxyu629 commented on scripts/effects/repair_effect.gd:
```diff
@@ -0,0 +1,5 @@
+class_name RepairEffect
+extends EffectBase
+
```
```suggestion


```

Cxyu629 commented on scripts/fault_scene/fault.gd:
```diff
@@ -0,0 +1,64 @@
+	state = State.FIXED
+
+func get_time_progress_ratio() -> float:
+	return 1 - (reboot_timer.time_left / reboot_timer.wait_time)
```
```suggestion

```

Cxyu629 commented on scripts/fault_scene/visuals.gd:
```diff
@@ -0,0 +1,62 @@
+			play_sabotaged()
+		
+		[Fault.State.REBOOTING, Fault.State.FIXED]:
+			play_fixed()
```
```suggestion
func emit_disappear_finished():
	disappear_finished.emit()

```

Cxyu629 commented on scripts/fault_scene/fault.gd:
```diff
@@ -0,0 +1,64 @@
+	if state == State.SABOTAGED:
+		if(value >= repair_target):
+			state = State.REBOOTING
+		repair_progress.emit(value / repair_target)
```
```suggestion
		repair_progressed.emit(value / repair_target)
```

Cxyu629 commented on scripts/fault_scene/fault.gd:
```diff
@@ -0,0 +1,64 @@
+		[_, _]:
+			assert(false, "Invalid state change from %s to %s." % [State.find_key(from), State.find_key(to)])
+
+signal repair_progress(progress: float)
```
```suggestion

```

Cxyu629 commented on scripts/properties/melee_cooldown_prop.gd:
```diff
@@ -25,4 +25,4 @@ func can_melee() -> bool:
 	return _timer.is_stopped()
 
```
```suggestion


```

Cxyu629 commented on scripts/fault_scene/visuals.gd:
```diff
@@ -0,0 +1,62 @@
+	repair_bar.visible = false
+
+func exit():
+	disappeared.emit()
```
```suggestion

```

Cxyu629 reviewed (changes requested):
Requesting formatting changes + rename `exit` in `fault_scene/visuals.gd` to `emit_disappear_finished` and signal `disappeared` to `disappear_finished`.

deltaMinor committed:
feat: add power fault to enemy targets

deltaMinor committed:
feat: implement random power fault spawning

- yet to implement mechanism to despawn power fault if far away from the player

deltaMinor committed:
feat: implemented clearing of faraway faults

deltaMinor committed:
fix: fix power faults despawning incorrectly

deltaMinor committed:
Merge branch 'dev' into feat/6-Implement-Power-Fault

deltaMinor committed:
fix: update project version

deltaMinor committed:
Revert "fix: update project version"

This reverts commit f72a94a6f5bd7b33a6ed31b25b81316a0bf1b6b1.

Cxyu629 committed:
style: fix formatting

Cxyu629 committed:
fix: add effect to player's melee

Additionally fixed broken debug feature due to commits in current branch

Cxyu629 committed:
Merge branch 'feat/6-Implement-Power-Fault' of https://github.com/Crash-Out-Engine/NUS-Orbital-2025 into feat/6-Implement-Power-Fault

Cxyu629 committed:
Merge branch 'dev' into feat/6-Implement-Power-Fault
